### PR TITLE
PIM-5664 : Add command for purge job executions

### PR DESCRIPTION
**Pull request by momoss**
Add command for purge job executions

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | N
| Added Behats                      | N
| Changelog updated                 | Y
| Review and 2 GTM                  | Y
| Micro Demo to the PO (Story only) | N
| Migration script                  | N
| Tech Doc                          | Y

### CODE REVIEW

File: src/Akeneo/Bundle/BatchBundle/Command/PurgeJobExecutionCommand.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class PurgeJobExecutionCommand extends ContainerAwareCommand
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function configure()
+    {
+        $this->setName('akeneo:batch:purge-job-execution');
```
Comment (juliensnz) [inline src/Akeneo/Bundle/BatchBundle/Command/PurgeJobExecutionCommand.php]: please add a nice description, with help and so on :)

File: src/Akeneo/Bundle/BatchBundle/Command/PurgeJobExecutionCommand.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $jobsExecutions = $this->getJobExecutionRepository()->getExecutionsOlderThan(
+            $this->getNumberOfDaysParameters()
```
Comment (juliensnz) [inline src/Akeneo/Bundle/BatchBundle/Command/PurgeJobExecutionCommand.php]: should be a command parameter, not a conf

File: src/Akeneo/Bundle/BatchBundle/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -27,6 +27,10 @@ public function getConfigTreeBuilder()
             ->children()
                 ->booleanNode('enable_mail_notification')->defaultFalse()->end()
                 ->scalarNode('sender_email')->defaultValue('mailer@bap.com')->end()
+                ->arrayNode('purge_job_executions')
```
Comment (juliensnz) [inline src/Akeneo/Bundle/BatchBundle/DependencyInjection/Configuration.php]: we should not add a configuration for that

Comment (jjanvier) [inline src/Akeneo/Bundle/BatchBundle/DependencyInjection/Configuration.php]: useless imho, should be an option of the command

File: src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php [hunk 4]
```diff
@@ -101,4 +101,23 @@ public function createJobTrackerDatagridQueryBuilder()
 
         return $qb;
     }
+
+    /**
+     * @param integer $nbOfDays
+     *
+     * @return array
+     */
+    public function getExecutionsOlderThan($nbOfDays)
```
Comment (juliensnz) [inline src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php]: $days

File: src/Pim/Bundle/EnrichBundle/Resources/config/bundles/akeneo_batch.yml [hunk 5]
```diff
@@ -1,3 +1,5 @@
 akeneo_batch:
     sender_email:             'mailer@akeneo.com'
     enable_mail_notification: true
+    purge_job_executions:
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Resources/config/bundles/akeneo_batch.yml]: useless imho, should be an option of the command

File: src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php (hunk 4 above)
Comment (juliensnz) [inline src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php]: findJobExecutionsOlderThan

File: src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php [hunk 6]
```diff
@@ -101,4 +101,23 @@ public function createJobTrackerDatagridQueryBuilder()
...
+    {
+        $qb = $this->createQueryBuilder('e');
+
+        $today = new \DateTime();
+        $today->sub(new \DateInterval(sprintf('P%sD', $nbOfDays)));
+
+        $qb->where(
+            $qb->expr()->lt('e.endTime', sprintf("'%s'", $today->format('Y-m-d H:i:s')))
```
Comment (juliensnz) [inline src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php]: please use a constant for the date format

**Comment by juliensnz** [comment]
GTM after comment fixes

File: src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php [hunk 7]
```diff
@@ -101,4 +101,23 @@ public function createJobTrackerDatagridQueryBuilder()
...
+     * @param integer $nbOfDays
+     *
+     * @return array
+     */
+    public function getExecutionsOlderThan($nbOfDays)
+    {
+        $qb = $this->createQueryBuilder('e');
+
+        $today = new \DateTime();
+        $today->sub(new \DateInterval(sprintf('P%sD', $nbOfDays)));
```
Comment (skeleton) [inline src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php]: `$today->modify(sprintf('- %d days', $nbOfDays))` is more readable IMO

File: src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php (hunk 6 above)
Comment (skeleton) [inline src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php]: not sure $today is the good naming :)

Comment (skeleton) [inline src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php]: why do you add a sprintf ?

File: src/Akeneo/Bundle/BatchBundle/Command/PurgeJobExecutionCommand.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace Akeneo\Bundle\BatchBundle\Command;
+
+use Pim\Bundle\ImportExportBundle\Entity\Repository\JobExecutionRepository;
```
Comment (skeleton) [inline src/Akeneo/Bundle/BatchBundle/Command/PurgeJobExecutionCommand.php]: no. you cannot have a reference to the PIM inside Akeneo namespace

File: src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php [hunk 9]
```diff
@@ -101,4 +103,31 @@ public function createJobTrackerDatagridQueryBuilder()
...
+
+        $today = new \DateTime();
+        $today->sub(new \DateInterval(sprintf('P%sD', $days)));
+
+        $qb->where(
+            $qb->expr()->lt('je.endTime', sprintf("'%s'", $today->format(self::DATETIME_FORMAT)))
+        );
+
+        $subQb = $this->createQueryBuilder('jesub');
+        $subQb->select('MAX(jesub.id)');
```
Comment (skeleton) [inline src/Pim/Bundle/ImportExportBundle/Entity/Repository/JobExecutionRepository.php]: not enough, as the last job executed can be a fail, you need to add a condition to have only jobExecution with status = COMPLETED

File: src/Akeneo/Bundle/BatchBundle/Job/DoctrineJobRepository.php [hunk 10]
```diff
@@ -151,4 +153,46 @@ public function getLastJobExecution(JobInstance $jobInstance, $status)
             ->getQuery()
             ->getOneOrNullResult();
     }
+
+    /**
+     * @param integer $days
+     *
+     * @return array
+     */
+    public function findPurgeables($days)
```
Comment (skeleton) [inline src/Akeneo/Bundle/BatchBundle/Job/DoctrineJobRepository.php]: add these 2 new methods on JobRepositoryInterface. 
and don't forget to change comments with `{@inheritdoc}`

File: src/Akeneo/Bundle/BatchBundle/Resources/config/removers.yml [hunk 11]
```diff
@@ -1,8 +1,16 @@
...
     akeneo_batch.remover.job_instance:
-        class: %akeneo_storage_utils.remover.base.class%
+        class: '%akeneo_storage_utils.remover.base.class%'
         arguments:
             - '@doctrine.orm.entity_manager'
             - '@akeneo_storage_utils.remover.base_options_resolver'
             - '@event_dispatcher'
             - '%akeneo_batch.entity.job_instance.class%'
+
+    akeneo_batch.remover.job_execution:
```
Comment (skeleton) [inline src/Akeneo/Bundle/BatchBundle/Resources/config/removers.yml]: still used ?

**Comment by skeleton** [comment]
GTM

**Pull request merged**